File: src/main/kotlin/com/horsey/scraper/Credentials.kt
package com.horsey.scraper

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.nio.file.Files
import java.nio.file.LinkOption
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.attribute.PosixFileAttributeView

data class Credentials(
    val username: String,
    val password: String,
    val appKey: String,
)

/**
 * Pure parser. Accepts a JSON object with string fields `username`,
 * `password`, `appKey`. Extra fields are ignored. Missing fields throw
 * `IllegalArgumentException` listing every offender in one message.
 */
fun parseCredentials(json: String): Credentials {
    val root: JsonObject = try {
        JsonParser.parseString(json).asJsonObject
    } catch (e: Exception) {
        throw IllegalArgumentException("credentials JSON is not a valid object: ${e.message}")
    }
    val missing = mutableListOf<String>()
    fun stringOrMiss(key: String): String? {
        val el = root.get(key)
        if (el == null || !el.isJsonPrimitive || !el.asJsonPrimitive.isString) {
            missing += key; return null
        }
        return el.asString
    }
    val username = stringOrMiss("username")
    val password = stringOrMiss("password")
    val appKey   = stringOrMiss("appKey")
    require(missing.isEmpty()) {
        "credentials JSON missing or non-string fields: ${missing.joinToString(",")}"
    }
    return Credentials(username!!, password!!, appKey!!)
}

/** Default path: `~/.horsey-scraper/credentials.json`. */
fun defaultCredentialsPath(): Path =
    Paths.get(System.getProperty("user.home"), ".horsey-scraper", "credentials.json")

/**
 * Reads and parses the credentials file at [path]. Errors with the path
 * embedded for easy debugging. If the file mode is wider than `0600` on a
 * POSIX filesystem, prints a single warning to stderr and continues.
 */
fun loadCredentials(path: Path): Credentials {
    if (!Files.exists(path)) {
        throw IllegalArgumentException("credentials file not found: $path")
    }
    warnIfWorldReadable(path)
    val json = try {
        Files.readString(path)
    } catch (e: Exception) {
        throw IllegalArgumentException("failed to read $path: ${e.message}")
    }
    return parseCredentials(json)
}

private fun warnIfWorldReadable(path: Path) {
    val view = Files.getFileAttributeView(
        path, PosixFileAttributeView::class.java, LinkOption.NOFOLLOW_LINKS
    ) ?: return
    val perms = view.readAttributes().permissions()
    val tooOpen = perms.any { it.name.startsWith("GROUP_") || it.name.startsWith("OTHERS_") }
    if (tooOpen) {
        System.err.println("Warning: $path is readable by group/others; recommend `chmod 600`.")
    }
}

File: src/betfair_scraper/credentials.py
"""Load and validate ~/.horsey-scraper/credentials.json. Port of Credentials.kt."""

from __future__ import annotations

import json
import os
import sys
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class Credentials:
    username: str
    password: str
    app_key: str


def parse_credentials(text: str) -> Credentials:
    """JSON object with string fields username, password, appKey. Extra
    fields ignored. Missing/non-string fields → ValueError listing all."""
    try:
        root = json.loads(text)
        if not isinstance(root, dict):
            raise ValueError("not an object")
    except ValueError as e:
        raise ValueError(f"credentials JSON is not a valid object: {e}")
    missing: list[str] = []

    def s(key: str) -> "str | None":
        v = root.get(key)
        if not isinstance(v, str):
            missing.append(key)
            return None
        return v

    username, password, app_key = s("username"), s("password"), s("appKey")
    if missing:
        raise ValueError(
            f"credentials JSON missing or non-string fields: {','.join(missing)}"
        )
    return Credentials(username, password, app_key)


def default_credentials_path() -> Path:
    return Path.home() / ".horsey-scraper" / "credentials.json"


def load_credentials(path: Path | str) -> Credentials:
    path = Path(path)
    if not path.exists():
        raise ValueError(f"credentials file not found: {path}")
    _warn_if_world_readable(path)
    try:
        text = path.read_text()
    except OSError as e:
        raise ValueError(f"failed to read {path}: {e}")
    return parse_credentials(text)


def _warn_if_world_readable(path: Path) -> None:
    try:
        mode = os.stat(path).st_mode
    except OSError:
        return
    if mode & 0o077:
        print(
            f"Warning: {path} is readable by group/others; recommend `chmod 600`.",
            file=sys.stderr,
        )
